File: backend/src/main/java/com/smvita/computerseekho/entity/Payment.java
package com.smvita.computerseekho.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.math.BigDecimal;
import java.time.LocalDate;

/**
 * One installment against an enrolment.
 *
 * The schema stays general (installment_number / total_installments) while
 * the business rule fixes total_installments at 2 system-wide — KB §3.2,
 * resolved in Turn 5. Keeping the column general means a future change of
 * policy is a service-layer edit, not a migration.
 */
@Entity
@Table(name = "payments")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Payment {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "payment_id")
    private Integer paymentId;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "student_id", nullable = false)
    private Student student;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "enrollment_id", nullable = false)
    private Enrollment enrollment;

    @Column(name = "amount", nullable = false, precision = 10, scale = 2)
    private BigDecimal amount;

    @Column(name = "installment_number", nullable = false)
    private Integer installmentNumber = 1;

    @Column(name = "total_installments", nullable = false)
    private Integer totalInstallments = 2;

    @Column(name = "payment_date", nullable = false)
    private LocalDate paymentDate = LocalDate.now();

    @Convert(converter = PaymentModeConverter.class)
    @Column(name = "payment_mode", nullable = false,
            columnDefinition = "ENUM('Cash','UPI','Card','Bank Transfer','Cheque')")
    private Mode paymentMode = Mode.Cash;

    @Enumerated(EnumType.STRING)
    @Column(name = "payment_status", nullable = false,
            columnDefinition = "ENUM('Success','Pending','Failed','Refunded')")
    private Status paymentStatus = Status.Success;

    @Column(name = "transaction_id", length = 100)
    private String transactionId;

    /** Unique, human-readable, printed on the PDF. Generated in PaymentService. */
    @Column(name = "receipt_no", nullable = false, unique = true, length = 100)
    private String receiptNo;

    @Column(name = "remarks", columnDefinition = "TEXT")
    private String remarks;

    public enum Mode {
        Cash, UPI, Card, BankTransfer, Cheque;

        public String toDbValue() {
            return this == BankTransfer ? "Bank Transfer" : this.name();
        }

        public static Mode fromDbValue(String value) {
            return "Bank Transfer".equals(value) ? BankTransfer : Mode.valueOf(value);
        }
    }

    public enum Status { Success, Pending, Failed, Refunded }
}

File: dotnet_backend/backend/Models/Payment.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace ComputerSeekho.Models;

/// <summary>
/// One fee installment.
///
/// ReceiptNo is unique and derived rather than sequential — see
/// FeeCalculator for why a running counter would be a race.
/// </summary>
[Table("payments")]
public class Payment
{
    [Key]
    [Column("payment_id")]
    public int PaymentId { get; set; }

    [Column("student_id")]
    public int StudentId { get; set; }

    [Column("enrollment_id")]
    public int EnrollmentId { get; set; }

    [Column("amount")]
    public decimal Amount { get; set; }

    [Column("installment_number")]
    public int InstallmentNumber { get; set; } = 1;

    /// <summary>Fixed at 2 system-wide (BRD section 3.2).</summary>
    [Column("total_installments")]
    public int TotalInstallments { get; set; } = 2;

    [Column("payment_date")]
    public DateOnly PaymentDate { get; set; } = DateOnly.FromDateTime(DateTime.Today);

    [Column("payment_mode")]
    public PaymentMode PaymentMode { get; set; } = PaymentMode.Cash;

    [Column("payment_status")]
    public PaymentStatus PaymentStatus { get; set; } = PaymentStatus.Success;

    [Column("transaction_id")]
    [MaxLength(100)]
    public string? TransactionId { get; set; }

    [Column("receipt_no")]
    [Required, MaxLength(100)]
    public string ReceiptNo { get; set; } = string.Empty;

    [Column("remarks")]
    public string? Remarks { get; set; }

    [ForeignKey(nameof(StudentId))]
    public Student? Student { get; set; }

    [ForeignKey(nameof(EnrollmentId))]
    public Enrollment? Enrollment { get; set; }
}
